Fix pulling of images within buildah

rhatdan opened the pull request:
Change references to Transfer to transfer to make it internal only.
It should be determined from the image specification and only determined
in one place.

Make buildah.Pull use registries.conf

Currently buildah pull does not resolve images based on registries.conf
This does not match the behaviour of buildah from or buildah bud

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

TomSweeneyRedHat commented:
Does this resolve #1316?

rhatdan commented:
Fixes  #1316

rhatdan commented on pull.go:
```diff
@@ -152,60 +152,60 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
+			if options.ReportWriter != nil {
+				options.ReportWriter.Write([]byte("Pulling " + name + "\n"))
+			}
+			ref, err := pullImage(ctx, options.Store, name, options, systemContext)
```
Yes.  
srcRef, _, err := resolveImage(ctx, systemContext, options.Store, boptions)
Actually pulls the image.  The dump-tags is just using the resolved first image name to pull  all of the tags.

TomSweeneyRedHat commented:
I'd like to take a deeper dive an run some test with this Tues morn.  But if others OK this before, then fine by me.

mtrmac commented on pull.go:
```diff
@@ -152,60 +152,60 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
 func Pull(ctx context.Context, imageName string, options PullOptions) error {
-	spec := imageName
 	systemContext := getSystemContext(options.SystemContext, options.SignaturePolicyPath)
-	srcRef, err := alltransports.ParseImageName(spec)
```
If I am reading `return []string{split[1]}, false, nil` in `ResolveName` correctly, this is now broken — it only works if `options.Transport` matches the transport in the input.

_That_ problem, in turn, is masked by e.g. `buildah from` parsing the input string and setting `options.Transport` to make this all work — but non-CLI callers could be tripped by this, and it’s a very fragile way to structure the code anyway.

mtrmac commented on pull.go:
```diff
@@ -152,60 +152,60 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
-			transport = transport + ":"
-		}
-		spec = transport + spec
-		srcRef2, err2 := alltransports.ParseImageName(spec)
```
(This is also broken, for the more exotic values of `Transport`, but I don’t quite mind `DefaultTransport` etc. going away, the after-transport-specification suffixes are fundamentally not substitutable.)

mtrmac commented on pull.go:
```diff
@@ -152,60 +152,60 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
+		boptions.ReportWriter = nil // Turns off logging output
 	}
-	var names []string
+	srcRef, _, err := resolveImage(ctx, systemContext, options.Store, boptions)
```
The reference returned by `resolveImage` does not refer to the source (e.g. a remote registry), but to the destination (c/storage).

_If_ the source is from a `docker.Transport`, the `DockerReference()` value should be the same, so in that case this is primarily a matter of variable naming; if the source is from some other source entirely, this code will pull the other tags from some completely unrelated docker/distribution registry, based on how the pull code decided to name the pulled image.

mtrmac commented on pull.go:
```diff
@@ -152,60 +152,60 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
 			return errors.New("Non-docker transport is currently not supported")
 		}
+
+		spec := transport + srcRef.DockerReference().Name()
```
Looking at `resolveName`, the transport value does not include the `:` separator unless it is `docker://` (or `DefaultTransport`, depending on how exactly it is hard-coded); it needs to be added.

Or, well, if you want to hard-code `docker.Transport`, just do so.

rhatdan replied:
I would like to get rid of this code altogether and get it into a function provided by containers/image written by you or @vrothberg.  But for now, I would just like to fix the issue of buildah pull not working like buildah from.

mtrmac replied:
\*shrug\* I don’t quite buy “this is an important feature and that justifies not getting the details right”, but in the end, I’m just pointing out breakage I am noticing; I am not a buildah maintainer and I won’t be approving the PR.

mtrmac replied:
(To be clear, I am not saying that consolidation of the `transport + name` logic into a function is a requirement of this PR; the goal here was getting the `":"` handling right, and consolidation was a side suggestion.)

TomSweeneyRedHat commented:
When an image is pulled, the imageID is not being printed with this patch as it had been prior.

```
# buildah images
# buildah pull docker.io/busybox
# buildah images
IMAGE NAME                                               IMAGE TAG            IMAGE ID             CREATED AT             SIZE
docker.io/library/busybox                                latest               3a093384ac30         Dec 31, 2018 20:21     1.42 MB

```

TomSweeneyRedHat commented:
Not only no imageId at the end, but no progress bars.

rhatdan commented:
Fixed imageid and progress problems.

rhatdan commented:
@TomSweeneyRedHat Could you merge this one.

TomSweeneyRedHat commented:
Code and testing LGTM, but I'd like to get a head nod from @mtrmac to make sure his concerns were addressed.  I'm not quite sure about his second comment in the old code in particular having been addressed.

rhatdan commented:
I would rather get his concerns addressed in a new effort to get this code rewritten into containers/image and then vendored into podman, Buildah and CRI-O to do it the same way in all packages.

TomSweeneyRedHat commented:
I was hoping @mtrmac would be able to check in by now, but I'll go ahead and merge now.  I agree the right place for this functionality in my book is containers/image and it looks like we're moving in that direction.

TomSweeneyRedHat commented:
@rh-atomic-bot r+

mtrmac commented:
> Code and testing LGTM, but I'd like to get a head nod from @mtrmac to make sure his concerns were addressed. I'm not quite sure about his second comment in the old code in particular having been addressed.

AFAICS all of those review comments are still applicable.

In my experience it’s pretty much never worth it to knowingly introduce new bugs in order to get a new feature, _especially_ if introducing those bugs is not at all essential for the feature (as is the case of 
https://github.com/containers/buildah/pull/1319/files#r253626657 , which needs about three more lines; the other comments may require notably more code). But, \*shrug\*, I realize that’s a bit extremist approach, and there may not be time or, well, interest, to avoid such situations — again, not my call, but that decision did depend on me being happy with the PR, I’m afraid I’m still not.

---

On a more conceptual level, does it even make much sense to `buildah pull --all-tags fedora`? Does that mean “I want all `fedora:*` tags from all registries on the search list”, or “I want all `fedora:*` tags from the first registry on the search list that contains a `fedora`, but I don’t care at all which registry it is”? The latter seems a tiny bit more sane, but _both_ seem, fundamentally, rather useless. When would anyone _ever_ want to use either?

So, maybe, at the highest level, there should be a separate `PullAllTags` + `Pull`, with `PullAllTags` not using the search path at all (and minimally modified), and `Pull` modified roughly like this PR? That would also nicely bypass the `":"` and the `srcRef` concerns (but not the other `alltransports.ParseImageName` semantics breakage).

OTOH doing that the easy way would cause `PullAllTags` to (continue to) use the `docker.io/library` normalization, which may be considered very undesirable.

mtrmac commented:
> I would rather get his concerns addressed in a new effort to get this code rewritten into containers/image and then vendored into podman, Buildah and CRI-O to do it the same way in all packages.

(BTW don’t expect miracles from this — as https://github.com/containers/libpod/blob/9c1b08fd79012c12d478edb13f6057a0414762db/libpod/image/parts.go#L54 shows, rewrites in c/image won’t free libpod/buildah from making decisions about semantics of the input; if anything, it would make it less avoidable for them to make such decisions.)

rhatdan commented:
@mtrmac @nalind 
I reworked this PR Heavily to remove the handling of Transports from the clients, and moved it into the library.

I made the Transports private to the library to prevent the client passing in a transport different then the image.

Transport=docker
ImageName=ociarchive:/tmp/foobar

Also added OCI As a valid format.

util.ResolveImage now returns the image name without the transport as well as the transport.

TomSweeneyRedHat commented on docs/buildah-pull.md:
```diff
@@ -25,6 +25,9 @@ Multiple transports are supported:
   An image _docker-reference_ stored in the docker daemon's internal storage.  _docker-reference_ must include either a tag or a digest.  Alternatively, when reading images, the format can also be docker-daemon:algo:digest (an image ID).
 
+  **oci:**_path_**:**_tag_**
+  An image tag in a directory compliant with "Open Container Image Layout Specification" at path.
```
nit, s/path/_path_/

TomSweeneyRedHat replied:
Ugh love GitHub auto md sometimes

`s/path/_path_/`

TomSweeneyRedHat commented:
@nalind and/or @vrothberg PTAL

TomSweeneyRedHat commented on buildah.go:
```diff
@@ -340,7 +340,7 @@ type BuilderOptions struct {
 	// needs to be pulled and the image name alone, or the image name and
 	// the registry together, can not be resolved to a reference to a
 	// source image.  No separator is implicitly added.
-	Transport string
```
Just wondering why you switched this to lower case?  Also in pull.go.  I understand the scoping, but none of the other variables in these structs are lowercase.  In some of the code we have a number of variables named transport(s) being used and having this upper case helps it to stand out better.

rhatdan replied:
I don't want callers to use this field.  We use it to pass the information around withing buildah library, but specifying transport as well as imagename can cause conflicts.  
Bottom line, I only want transport derived in one place and from ImageName.  Before we had it being derived in different places,and not the same way.

mtrmac commented on pull.go:
```diff
@@ -215,13 +222,10 @@ func pullImage(ctx context.Context, store storage.Store, imageName string, optio
 	spec := imageName
 	srcRef, err := alltransports.ParseImageName(spec)
 	if err != nil {
```
The two callers of `pullImage` now explicitly know whether the input should be readable by `ParseImageName` or whether it needs the transport added (and using the other path can misinterpret the input); it would be nice to make this unambiguous.

`pullAndFindImage` AFAICS currently always submits input input with the “concatenate `options.transport`+`imageName`” format, while the `AllTags` path in `Pull` always submits input with the “use `ParseImageName(imageName)` semantics — so, the concatenation code could be moved to `pullAndFindImage` (or maybe even higher to `resolveImage`, and consolidated with the other part there), and `pullImage` could expect input in the `ParseImageName` format.

(This is a pre-existing situation, so non-blocking; the changes to `ResolveName` in this PR now allow making this unambiguous.)

rhatdan replied:
Lets handle this in a separate PR

rhatdan commented on buildah.go:
```diff
@@ -340,7 +340,7 @@ type BuilderOptions struct {
 	// the registry together, can not be resolved to a reference to a
 	// source image.  No separator is implicitly added.
-	Transport string
+	transport string
```
Removing.

rhatdan commented on cmd/buildah/pull.go:
```diff
@@ -80,27 +78,14 @@ func pullCmd(c *cobra.Command, args []string, iopts pullResults) error {
-	if len(arr) == 2 {
-		if iopts.allTags {
-			return errors.Errorf("tag can't be used with --all-tags")
-		}
```
This is tough to check, and seem fairly arbitrary.  If I state I want to pull --alltags on alpine:latest, I don't see this as a big conflict.

rhatdan commented on imagebuildah/build.go:
```diff
@@ -582,7 +577,6 @@ func NewExecutor(store storage.Store, options BuildOptions) (*Executor, error) {
 		contextDir:                     options.ContextDirectory,
 		pullPolicy:                     options.PullPolicy,
 		registry:                       options.Registry,
-		transport:                      options.Transport,
```
Removed

rhatdan commented on util/util.go:
```diff
@@ -70,37 +48,39 @@ var (
 // and a boolean which is true iff 1) the list of search registries was used, and 2) it was empty.
```
Added Docs.

rhatdan commented on util/util.go:
```diff
@@ -70,37 +48,39 @@ var (
 			// It's a truncated version of the ID of an image that's present in local storage;
 			// we need only expand the ID.
-			return []string{img.ID}, false, nil
+			return []string{img.ID}, DefaultTransport, false, nil
```
Changed to ""

rhatdan commented on util/util.go:
```diff
@@ -70,37 +48,39 @@ var (
-		if _, ok := Transports[split[0]]; ok {
-			return []string{split[1]}, false, nil
+	if strings.HasPrefix(name, DefaultTransport) {
+		name = strings.TrimPrefix(name, DefaultTransport)
```
Fixed.

rhatdan commented on new.go:
```diff
@@ -33,7 +33,7 @@ func pullAndFindImage(ctx context.Context, store storage.Store, imageName string
 		Store:         store,
 		SystemContext: options.SystemContext,
-		Transport:     options.Transport,
+		transport:     options.transport,
```
Made into a parameter.

rhatdan commented on new.go:
```diff
@@ -250,21 +253,22 @@ func findUnusedContainer(name string, containers []storage.Container) string {
-	if options.Transport == "" {
-		options.Transport = util.DefaultTransport
+	if options.transport == "" {
+		options.transport = util.DefaultTransport
```
Removed

rhatdan commented on pull.go:
```diff
@@ -152,60 +162,57 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
 	}
-	var names []string
+
+	fmt.Printf("%s\n", img.ID)
```
Fixed.

rhatdan commented on pull.go:
```diff
@@ -152,60 +162,57 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
-		options.ReportWriter = nil // Turns off logging output
+		boptions.ReportWriter = nil // Turns off logging output
+	}
+	srcRef, transport, img, err := resolveImage(ctx, systemContext, options.Store, boptions)
```
If I don't test for the conflict of TAG versus all-tag, then the user must specify a legal tag if the :latest does not exist to do a pull-all. 

Is that adequate for now.

mtrmac replied:
At the very least, `s/srcRef/destRef/` or `s/srcRef/storageRef/` to not be confusing about the semantics of that value, please.

---

WRT requiring `:latest` to exist:

Well, it’s a breaking change AFAICS; this used to work. And forcing users to specify a tag would make reintroducing the “`--all-tags` with a tag is invalid” error _another_ breaking change.

\*shrug\* I don’t quite understand the utility of `--all-tags` (let alone search+`--all-tags` together) in the first place, so I guess it is acceptable.

rhatdan commented on pull.go:
```diff
@@ -152,60 +162,57 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
-			errs = multierror.Append(errs, err)
-			continue
+			if options.ReportWriter != nil {
+				options.ReportWriter.Write([]byte("Pulling " + name + "\n"))
```
Removed

mtrmac commented on util/util.go:
```diff
@@ -43,64 +40,47 @@ var (
+	if strings.HasPrefix(name, DefaultTransport) {
+		return []string{strings.TrimPrefix(name, DefaultTransport)}, DefaultTransport, false, nil
+	}
 	// If the image includes a transport's name as a prefix, use it as-is.
```
Tiny nit: This comment would better fit above the `DefaultTransport` case.

rhatdan replied:
Moved

mtrmac commented on pull.go:
```diff
@@ -152,76 +154,71 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
+			if options.ReportWriter != nil {
+				options.ReportWriter.Write([]byte("Pulling " + name + "\n"))
+			}
+			ref, err := pullImage(ctx, options.Store, transport, name, options, systemContext)
```
Non-blocking: 

For consistency with the `pullAndFindImage` caller, it would be nice if `name` here did not start with `transport`:
```go
repoName := whateverSrcRefIsRenamedTo.DockerReference().Name()
srcRef, err := alltransports.ParseImageName(transport+repoName)
// or srcRef, err := c/image/docker.NewReference(srcRef.DockerReference())
…
name = repoName + ":" + tag
```
or so.

rhatdan commented:
@mtrmac Ready to merge.  I will open another PR to further clean this up and eventually get it into podman and maybe add the functions to containers/image

@rh-atomic-bot r+

mtrmac commented:
CK. Thanks!

mtrmac commented:
ACK, that is.
